framework/src/meshgenerators/MeshDiagnosticsGenerator.C

```diff
@@ -1252,6 +1262,90 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
                  num_negative_side_qp_jacobians);
 }
 
+void
+MeshDiagnosticsGenerator::checkNonMatchingEdges(const std::unique_ptr<MeshBase> & mesh) const
+{
+  /*Algorithm Overview
+    1)Prechecks
+      a)This algorithn only works for 3D so check for that first
+    2)Loop
+      a)Loop through every element
+      b)For each element get the edges associated with it
+      c)For each edge check overlap with any edges of nearby elements
+      d)Have check to make sure the same pair of edges are not being tested twice for overlap
+    3)Overlap check
+      a)Shortest line that connects both lines is perpendicular to both lines
+      b)A good overview of the math for finding intersecting lines can be found
+    here->paulbourke.net/geometry/pointlineplane/
+  */
+  if (mesh->mesh_dimension() != 3)
+    mooseError("The edge intersection algorithm only works with 3D meshes");
+  if (!mesh->is_serial())
+    mooseError("Only serialized/replicated meshes are supported");
+  unsigned int num_intersecting_edges = 0;
+  std::vector<Node> checked_edges;
+  for (auto elem : mesh->active_element_ptr_range())
+  {
+    std::vector<std::unique_ptr<Elem>> elem_edges(elem->n_edges());
+    for (auto i : elem->edge_index_range())
+      elem_edges[i] = elem->build_edge_ptr(i);
+    for (auto other_elem : mesh->active_element_ptr_range())
+    {
+      // If they're the same element then there's no need to check for overlap
+      if (elem == other_elem)
+      {
+        continue;
+      }
+      // Get bounding boxes for both elements. If the bounding boxes don't intersect then no edges
+      // will either
+      auto boundingBox1 = elem->loose_bounding_box();
+      auto boundingBox2 = other_elem->loose_bounding_box();
+      if (!(boundingBox1.intersects(boundingBox2)))
+      {
+        continue;
+      }
+      std::vector<std::unique_ptr<Elem>> other_edges(other_elem->n_edges());
+      for (auto j : other_elem->edge_index_range())
+        other_edges[j] = other_elem->build_edge_ptr(j);
+      for (auto & edge : elem_edges)
+      {
+        for (auto & other_edge : other_edges)
```

Comment:
So maybe there is a better way, but I used the reference because elem->build_edge_ptr() creates a unique pointer to the edge, so when I don't take the reference I get an error because it's trying to create a copy of a unique_ptr. If you know of a better way around this let me know!

Reply:
ah no this is fine. I didnt see they were unique_ptrs